File: scripts/projections/skills.py
"""data.skills/<skill_key>.json + _index.json projections — per DATA_ARCHITECTURE.md §6.5 / §6.6.

Status: Future (skipped by default in build_data.py)
Consumer: future "find jobs by skill" page
Shape:
    _index.json: {"skills": [{"key", "label_ja"}]}
    <key>.json: {"skill_key", "label_ja", "occupations": [{"id", "name_ja", "score"}]}
Size target: < 15 KB per per-skill file

`occupations` is sorted by score descending; only occupations with non-null skills block participate.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from scripts.lib.indexes import Indexes


def build(indexes: "Indexes", dist_root: Path) -> dict:
    out_dir = dist_root / "data.skills"
    out_dir.mkdir(parents=True, exist_ok=True)

    skills_labels = indexes.labels_by_dim.get("skills", {})

    # Per-skill ranking files
    written = 0
    for skill_key, label in skills_labels.items():
        ranked: list[dict] = []
        for occ_id, occ in indexes.occ_by_id.items():
            if occ.skills is None:
                continue
            score = occ.skills.get(skill_key)
            if score is None:
                continue
            ranked.append({"id": occ_id, "name_ja": occ.title_ja, "score": score})
        ranked.sort(key=lambda x: x["score"], reverse=True)

        payload = {
            "skill_key": skill_key,
            "label_ja": label.ja,
            "occupations": ranked,
        }
        out = out_dir / f"{skill_key}.json"
        out.write_text(json.dumps(payload, ensure_ascii=False, separators=(",", ":")) + "\n", encoding="utf-8")
        written += 1

    # _index.json
    index_payload = {
        "schema_version": "1.0",
        "skills": [
            {"key": k, "label_ja": v.ja}
            for k, v in skills_labels.items()
        ],
    }
    index_path = out_dir / "_index.json"
    index_path.write_text(json.dumps(index_payload, ensure_ascii=False, separators=(",", ":")) + "\n", encoding="utf-8")

    return {"dir": out_dir, "skill_files": written, "index": index_path}

File: src/data/projections/skills.ts
/**
 * data.skills/<skill_key>.json + _index.json projections — per docs/DATA_ARCHITECTURE.md §6.5 / §6.6.
 *
 * Status: Future (skipped by default)
 * Consumer: future "find jobs by skill" page
 *
 * Migrated from scripts/projections/skills.py.
 */
import { mkdir, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import type { Indexes } from '../lib/indexes.js';
import type { LabelEntry } from '../schema/labels.js';

export interface SkillsBuildResult {
  files: string[];
  dir: string;
  skillFiles: number;
  indexFile: string;
}

export async function buildSkills(
  indexes: Indexes,
  distRoot: string,
): Promise<SkillsBuildResult> {
  const outDir = join(distRoot, 'data.skills');
  await mkdir(outDir, { recursive: true });

  const skillsLabels = indexes.labelsByDim.get('skills') ?? new Map<string, LabelEntry>();

  const writtenFiles: string[] = [];

  for (const [skillKey, label] of skillsLabels) {
    const ranked: Array<{ id: number; name_ja: string; score: number }> = [];
    // Match Python: iterate `indexes.occ_by_id.items()` in insertion order.
    for (const [occId, occ] of indexes.occById) {
      if (occ.skills == null) continue;
      const score = occ.skills[skillKey];
      if (score == null) continue;
      ranked.push({ id: occId, name_ja: occ.title_ja, score });
    }
    // Sort by score desc, stable.
    ranked.sort((a, b) => b.score - a.score);

    const payload = {
      skill_key: skillKey,
      label_ja: label.ja,
      occupations: ranked,
    };
    const filePath = join(outDir, `${skillKey}.json`);
    await writeFile(
      filePath,
      JSON.stringify(payload) + '\n',
      'utf-8',
    );
    writtenFiles.push(filePath);
  }

  // _index.json
  const indexSkills: Array<{ key: string; label_ja: string }> = [];
  for (const [key, value] of skillsLabels) {
    indexSkills.push({ key, label_ja: value.ja });
  }
  const indexPayload = {
    schema_version: '1.0',
    skills: indexSkills,
  };
  const indexPath = join(outDir, '_index.json');
  await writeFile(
    indexPath,
    JSON.stringify(indexPayload) + '\n',
    'utf-8',
  );

  return {
    files: [...writtenFiles, indexPath],
    dir: outDir,
    skillFiles: writtenFiles.length,
    indexFile: indexPath,
  };
}
